### PR TITLE
Added search functionality [Resolves #7]

### LINKED ISSUE #7: [FEATURE] Tags/users search
**Describe the solution you'd like**
For my use case a feature to search for users or hashtags is needed to build an autocomplete field. 

A function should receive a search query and return an array of users and hashtags were found with avatar, description for user profiles and number of posts for hashtags.

The data will be taken from the search input on Instagram page:
Screen Shot 2019-06-28 at 0 40 23

**Proposal**

A public method with a following signature:

```ts
search: async (query: string, options: Instamancer.IOptions): Promise<Response>
```

I can implement this feature, if you will accept it. But for now I can see that it can conflict with the ideology of the project, because we need to fire DOM events first to capture the response. Maybe you can have an idea how we can implement this?

### PR DESCRIPTION
**Pull request by goooseman**
A new functionality to search for users/hashtags and places has been added.

### CODE REVIEW

**Comment by goooseman** [comment]
I see that AppVeyor build is failing, but this a new tool for me.

I will be happy to resolve the problem, if I will get any clue what to do.

**Comment by goooseman** [comment]
I've just wanted to open another PR to add some code quality tools like tslint or prettier, because I was sure project does not have any.

But now I see there is this Codacy tool.

The problem is that I can not see the result of the listing until I push a PR.
And I can not fix the issues automatically (like tslint does).

I will fix the issues now and update the PR.

**Comment by ScriptSmith** [comment]
Hi again, thanks for contributions!

I'll review them all properly later today, but for now I'll just let you know to ignore the AppVeyor checks as I thought I'd removed all traces of it from this project, but it appears I've left in the webhooks, so I apologize for the confusion.

File: src/api/search.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
+import {IOptions} from "./api";
+import {Instagram} from "./instagram";
+
+export interface ISearchResultUser {
```
[the remark this answers is not in the document]
Comment (goooseman) [reply src/api/search.ts]: I will solve it on a next branch

File: src/api/search.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        this.searchQuery = query;
+    }
+
+    public async get() {
+        if (!this.started) {
+            await this.start();
+        }
+        await this.page.click("input[type='text']");
+        await this.page.keyboard.type(this.searchQuery);
+        await this.page.waitForRequest((req) => this.matchURL(req.url()));
```
[the remark this answers is not in the document]
Comment (goooseman) [reply src/api/search.ts]: The delay between key presses is 0 [by default](https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#keyboardtypetext-options).

So no only one request is sent.

**Comment by goooseman** [comment]
I will close this PR and will open a new one.

File: src/api/search.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (ScriptSmith) [reply src/api/search.ts]: Maybe it's platform dependent, but in my case there were many requests being sent when I first tested this, and now.

Here's a snippet which simplifies it:
```typescript
import {launch} from "puppeteer";

(async () => {
    const browser = await launch({
        devtools: true
    });
    const page = (await browser.pages())[0];
    await page.goto("https://instagram.com/explore/tags/puppies");

    console.log("Filter for XHR requests in the network tab")
    await new Promise((resolve) => {
        setTimeout(resolve, 10000)
    })
    console.log("Done");

    await page.click("input[type='text']"); 
    await page.keyboard.type("This is a long string of text. qwertyuiopasdfghjklzxcvbnm.", {delay: 0}); 
})()
```
On my end, 6 other requests are sent before the final one is. Obviously that's an exaggerated example, but I could get it to send multiple requests with strings less than 10 characters.

[the remark this answers is not in the document]
Comment (goooseman) [reply src/api/search.ts]: Ok, in the new branch I will create a test to be sure, that only 1 request is being sent for a long input like this one.

**Pull request closed without merging**